web-local/src/lib/svelte-query/actions.ts

```diff
@@ -70,3 +80,100 @@ export async function deleteFileArtifact(
     console.error(err);
   }
 }
+
+export interface CreateDashboardFromSourceRequest {
+  instanceId?: string;
+  sourceName?: string;
+}
+
+export interface CreateDashboardFromSourceResponse {
+  dashboardName: string;
+}
+
+export const useCreateDashboardFromSource = <
+  TError = RpcStatus,
+  TContext = unknown
+>(options?: {
+  mutation?: UseMutationOptions<
+    Awaited<Promise<CreateDashboardFromSourceResponse>>,
+    TError,
+    { data: CreateDashboardFromSourceRequest },
+    TContext
+  >;
+}) => {
+  const { mutation: mutationOptions } = options ?? {};
+
+  const mutationFn: MutationFunction<
+    Awaited<Promise<CreateDashboardFromSourceResponse>>,
+    { data: CreateDashboardFromSourceRequest }
+  > = async (props) => {
+    const { data } = props ?? {};
+
+    // first, create model from source
+
+    // not ideal that this doesn't come from the useQuery cache
+    const existingModelFiles = await runtimeServiceListFiles(data.instanceId, {
```

Comment:
Not a fan of duplicating fetching files and parsing. Will be hard to maintain. We should get rid of this tech debt ASAP. But fine for the release

Reply:
Another option I was considering was providing model & dashboard files as an input to the hook, e.g.:

```typescript
$: existingModelNames = useModelNames(instanceId);
$: existingDashboardNames = useDashboardNames(instanceId);

$createDashboardFromSourceMutation.mutate(
      {
        data: { instanceId, sourceName, existingModelNames, existingDashboardNames  },
      },
      ...
)
```

Advantages: we leverage the useQuery cache, and the mutation becomes more like a pure function.
Disadvantages: we have to call the useXNames hooks in every component where we use this mutation, and the mutation doesn't encapsulate all the data fetching logic it needs.

You like this approach? Pros/cons to each, happy to go either way.

Reply:
This might be better I feel. More explicit with dependant data. We had implicit usage in redux and even though it made code small it was less obvious.